src/core/interval.js

```diff
@@ -18,6 +18,34 @@ export type Interval = {|
   +endTimeMs: TimestampMs,
 |};
 
+/**
+ * An interval sequence is an array of intervals with the following guarantees:
+ * - Every interval has non-negative time spam (i.e. the end time is greater
+ *   than or equal to the start time).
+ * - Every interval except for the first starts at the same time that the previous interval ended.
+ */
+export opaque type IntervalSequence: $ReadOnlyArray<Interval> = $ReadOnlyArray<Interval>;
+
+export function intervalSequence(
+  intervals: $ReadOnlyArray<Interval>
+): IntervalSequence {
+  let lastEndTime = null;
+  for (const {startTimeMs, endTimeMs} of intervals) {
+    if (lastEndTime != null && startTimeMs !== lastEndTime) {
+      throw new Error(
+        `last interval ended at ${lastEndTime} but this interval starts at ${startTimeMs}`
+      );
+    }
+    if (endTimeMs < startTimeMs) {
+      throw new Error(
+        `negative length interval: ${endTimeMs} < ${startTimeMs}`
+      );
+    }
+    lastEndTime = endTimeMs;
+  }
+  return intervals;
```

Comment:
Unsound; this type-checks but fails:

```javascript
    it("resists attack", () => {
      const i0 = {startTimeMs: 0, endTimeMs: 100};
      const i1 = {startTimeMs: 100, endTimeMs: 200};
      const is = intervalSequence([i0, i1]);
      i1.startTimeMs = 300;
      intervalSequence((is: IntervalSequence)); // fails
    });
```

Deep-copy?

Reply:
Fixed.